rfcs/2021-12-20-10517-llvm-backend-for-vrl.md

```diff
@@ -0,0 +1,560 @@
+  %61 = load void (%116*, {}*, %74*, %135*)*, void (%116*, {}*, %74*, %135*)** %60, align 8, !invariant.load !1, !noalias !99611, !nonnull !1
+  call void %61(%116* noalias nocapture nonnull sret(%116) dereferenceable(24) %7, {}* nonnull align 1 %55, %74* noalias nonnull readonly align 8 dereferenceable(32) bitcast ([32 x i8]* @16149 to %74*), %135* noalias nocapture nonnull dereferenceable(40) %6) #104, !noalias !99608
+  call void @llvm.lifetime.end.p0i8(i64 40, i8* nonnull %58), !noalias !99611
+  %62 = getelementptr inbounds %116, %116* %7, i64 0, i32 0
+  %63 = load {}*, {}** %62, align 8, !noalias !99611
+  %64 = icmp eq {}* %63, null
+  %65 = bitcast {}* %63 to i8*
+  br i1 %64, label %71, label %66
+
+66:                                               ; preds = %51
+  %67 = getelementptr inbounds %116, %116* %7, i64 0, i32 1, i64 0
+  %68 = load i64, i64* %67, align 8, !noalias !99611
+  %69 = icmp eq i64 %68, 0
+  br i1 %69, label %71, label %70
+
+70:                                               ; preds = %66
+  call void @__rust_dealloc(i8* nonnull %65, i64 %68, i64 1) #104, !noalias !99608
+  br label %71
+
+71:                                               ; preds = %51, %66, %70
+  call void @llvm.lifetime.end.p0i8(i64 24, i8* nonnull %53), !noalias !99611
+  call void @llvm.lifetime.end.p0i8(i64 40, i8* nonnull %52), !noalias !99611
+  br label %47
+}
+```
+
+Note the batched stack allocations, inlining of function calls and consolidation of control flow.
+
+## Rationale
+
+As long as the single-core performance of VRL is the bottleneck of a topology, general performance improvements to VRL are extremely valuable as they equate to an equally sized performance improvement to the entire topology.
+
+We want to live up to the [performance guarantees](https://github.com/vectordotdev/vector/blob/f1404bea186ba83c4426a32bbef3f633c17cf4d2/website/cue/reference/remap/features/compilation.cue#L4-L8) outlined in VRL's list of features.
+
+Every VRL program benefits from the reduced runtime overhead, without us needing to optimize any specific use case. Consistent execution speed is important to build trust in the language.
+
+Being able to execute our log transformation DSL at speeds which would otherwise only be attainable by hand-writing Rust programs will strengthen a key value proposition of Vector: best-in-class performance.
+
+## Drawbacks
+
+By generating machine code via LLVM, we are no longer (largely) immune to memory violations. To reduce the error surface as much as possible, we employ industry practices like fuzz-testing VRL programs, static analysis via LLVM and rely on the Rust compiler for any non-trivial code fragments. That being said, memory safety is an illusion - with the difference that we are now exposing ourselves to being partly responsible for maintaining invariants instead of being able to defer to a third party for correctness. We still provide an inherently memory safe language to the user.
```

Comment:
This idea may not make sense, but: is there any way we can lean more on stdlib functions to reduce the surface area of code that is at risk of containing bugs?

Essentially, in my mind, if we're using a lot of Rust stub code -- essentially, more of those exported Rust functions that you showed in your examples -- to handle basic operations, and we rely on LLVM to aggressive inline/optimize those, then I think we should theoretically get the benefits of Rust itself while limiting how much "raw" LLVM IR we end up having to emit?

Reply:
> Essentially, in my mind, if we're using a lot of Rust stub code -- essentially, more of those exported Rust functions that you showed in your examples -- to handle basic operations, and we rely on LLVM to aggressive inline/optimize those, then I think we should theoretically get the benefits of Rust itself while limiting how much "raw" LLVM IR we end up having to emit?

This is a good summary of that we're trying to do. So your idea not only makes sense, it's the gist of this PR 😄

The idea is to compose the VRLs functionality entirely via the following LLVM instructions only:
- `alloca` stack allocations
- `br` (conditional and unconditional branching)
- `call`s to our Rust stubs
- `global`s for constants

That makes the implementation quite maintainable to all Rust programmers, even with only a superficial understanding of LLVM. And LLVM optimizations passes take care for the rest.

Reply:
Gotcha.  I likely missed that after reading through the RFC the first time, so I appreciate the clarification. 👍🏻

Reply:
No problem, thanks for asking. I think that's a good hint that I should put that summary explicitly into the RFC.

Reply:
Added that summary in e8ef3bd078255cd46ffdced5d0f40deb21b028b2.